tokens/usecase/tokens_usecase.go

```diff
@@ -349,26 +391,69 @@ func (t *tokensUseCase) getPricesForBaseDenom(ctx context.Context, baseDenom str
 }
 
 func (t *tokensUseCase) getChainScalingFactorMut(precision int) (osmomath.Dec, bool) {
-	result, ok := t.precisionScalingFactorMap[precision]
-	return result, ok
+	result, ok := t.precisionScalingFactorMap.Load(precision)
+	if !ok {
+		return osmomath.Dec{}, false
+	}
+
+	v, ok := result.(osmomath.Dec)
+	if !ok {
+		return osmomath.Dec{}, false
+	}
+
+	return v, true
+}
+
+// UpdateAssetsAtHeightInterval updates assets at configured height interval.
+// Internally, it calls LoadTokensFromChainRegistry as a goroutine.
+func (t *tokensUseCase) UpdateAssetsAtHeightInterval(height uint64) {
+	if height%uint64(t.updateAssetsHeightInterval) == 0 {
+		go func() {
+			if err := t.LoadTokensFromChainRegistry(); err != nil {
+				t.logger.Error("error loading tokens from chain registry", zap.Error(err))
+			}
+		}()
+	}
+}
+
+// LoadTokensFromChainRegistry loads tokens from the chain registry into TokensUsecase.
+// Tokens are loaded only if the contents of the chain registry file have changed.
+func (t *tokensUseCase) LoadTokensFromChainRegistry() error {
+	tokens, hash, err := GetTokensFromChainRegistry(t.chainRegistryAssetsFileURL)
+	if err != nil {
+		return err
+	}
+
+	if t.lastLoadTokensFromChainRegistryHash != hash {
+		t.LoadTokens(tokens)
+	}
+
+	return nil
 }
```

Comment:
I think we should also unit-test this method. We should define `getTokensFromChainRegistry` as a field on `tokensUseCase`.

In core logic, we can wire up the actual implementation.

In the test logic, we can monkey-patch a mock.

Let me know if you'd like to have more pointers on how to get this working

Reply:
Please see my comment on implementation: https://github.com/osmosis-labs/sqs/pull/323#discussion_r1665920048